### PR TITLE
Fix MulticasInterface SocketOption to work with IPv6 on Linux

### LINKED ISSUE #25525: SocketOptionName.MulticastInterface not supported with Linux IPv6
Set the multicast interface socket option on IPv6 socket causes a SocketException

```c#
using System;
using System.Net.Sockets;

namespace client
{
    class Program
    {
        static void Main(string[] args)
        {
            var socket = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp);
            socket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastInterface, 1);
        }
    }
}
```

When run this on Ubuntu Linux with dotnet 2.0.4 results in a SocketException

```
Unhandled Exception: System.Net.Sockets.SocketException: Operation not supported
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Int32 optionValue, Boolean silent)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Int32 optionValue)
   at client.Program.Main(String[] args) in /home/jose/client/Program.cs:line 11
```

[EDIT] Add C# syntax highlight by @karelz

### PR DESCRIPTION
**Pull request by pepone**
Fixes #25525

### CODE REVIEW

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 1]
```diff
@@ -159,6 +159,28 @@ public void MulticastInterface_Set_InvalidIndex_Throws()
             }
         }
 
+        [OuterLoop] // TODO: Issue #11345
+        [Fact]
+        public void MulticastInterface_Set_WithIPv6()
+        {
+            using (Socket s = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp))
+            {
+                s.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastInterface, IPAddress.HostToNetworkOrder(1));
```
Comment (tmds) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: `IPAddress.HostToNetworkOrder` should not be needed here:

For IPV6_MULTICAST_IF:
> The input value for setting this option is a 4-byte interface index of the desired outgoing interface in **host byte order**.

Whereas for IP_MULTICAST_IF:
> The input value for setting this option is a 4-byte IPv4 address in network byte order. This DWORD parameter can also be an interface index in **network byte order**.

**Comment by tmds** [comment]
Run the OuterLoop tests (tests have OuterLoop attribute):

@dotnet-bot test Outerloop Linux x64 Debug Build please
@dotnet-bot test Outerloop OSX x64 Debug Build please

**Comment by pepone** [comment]
Are Outerloop tests `./run-test.sh` on Linux or do I need to set an specific parameter

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs (hunk 1 above)
Comment (pepone) [reply src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: Push a fix for that

**Comment by tmds** [comment]
> Are Outerloop tests ./run-test.sh on Linux or do I need to set an specific parameter

I think you can specify the --outerloop option to ./run-tests.sh.
I'd remove the OuterLoop attribute from the tests locally because it will take less time to run the tests (than running all the OuterLoop tests).

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 2]
```diff
@@ -159,6 +159,28 @@ public void MulticastInterface_Set_InvalidIndex_Throws()
...
+
+        [OuterLoop] // TODO: Issue #11345
+        [Fact]
+        public void MulticastInterface_Set_InvalidIndexWithIPv6_Throws()
+        {
+            int interfaceIndex = 31415;
+            using (Socket s = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp))
+            {
+                Assert.Throws<SocketException>(() =>
+                                               s.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastInterface, interfaceIndex);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: This doesn't compile.  It's missing a `)`.

Comment (pepone) [reply src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: Yes I noticed this, I fixing it and will push with some extra fixes requires in pal_networking.c

**Comment by pepone** [comment]
I added a better test and fix TryGetPlatformSocketOption to allow this option to be set

 ```
        Stack Trace:
           /home/jose/corefx/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(5112,0): at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
           /home/jose/corefx/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(1328,0): at System.Net.Sockets.Socket.SendTo(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, EndPoint remoteEP)
           /home/jose/corefx/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(1358,0): at System.Net.Sockets.Socket.SendTo(Byte[] buffer, EndPoint remoteEP)
           /home/jose/corefx/src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs(209,0): at System.Net.Sockets.Tests.SocketOptionNameTest.<MulticastInterface_Set_WithIPv6Helper>d__12.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
           /home/jose/corefx/src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs(182,0): at System.Net.Sockets.Tests.SocketOptionNameTest.<MulticastInterface_Set_WithIPv6>d__11.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
     System.Net.Sockets.Tests.SocketOptionNameTest.MulticastInterface_Set_AnyInterfaceWithIPv6_Succeeds [FAIL]
```

I can not set the option but trying to send an UDP package with the option set is failing, is not clear why, any ideas?

**Comment by stephentoub** [comment]
cc: @tmds

**Comment by tmds** [comment]
@pepone does the log have the exception message too? that may give some insights into why the send is failing.

**Comment by pepone** [comment]
@tmds the exception message is  `System.Net.Sockets.SocketException : Network is unreachable`

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 3]
```diff
@@ -159,6 +159,77 @@ public void MulticastInterface_Set_InvalidIndex_Throws()
...
+            using (Socket receiveSocket = CreateBoundUdpIPv6Socket(out port),
+                          sendSocket = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp))
+            {
+                receiveSocket.ReceiveTimeout = 1000;
+                receiveSocket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.AddMembership, new IPv6MulticastOption(multicastAddress, interfaceIndex));
+
+                // https://github.com/Microsoft/BashOnWindows/issues/990
+                if (!PlatformDetection.IsWindowsSubsystemForLinux)
+                {
+                    sendSocket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastInterface, 1);
```
Comment (tmds) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: `1` should be `interfaceIndex`.

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 4]
```diff
@@ -159,6 +159,77 @@ public void MulticastInterface_Set_InvalidIndex_Throws()
...
+            {
+                return; // [ActiveIssue(24008)]
+            }
+
+            await MulticastInterface_Set_WithIPv6Helper(1);
+        }
+
+        private async Task MulticastInterface_Set_WithIPv6Helper(int interfaceIndex)
+        {
+            IPAddress multicastAddress = IPAddress.Parse("ff15::1:1");
```
Comment (tmds) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: Let's reduce the scope of this address: `ff11`.

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 5]
```diff
@@ -224,6 +295,21 @@ private static Socket CreateBoundUdpSocket(out int localPort)
...
 
+        // Create an Udp Socket and binds it to an available port
+        private static Socket CreateBoundUdpIPv6Socket(out int localPort)
+        {
+            Socket receiveSocket = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp);
+
+            // sending a message will bind the socket to an available port
+            string sendMessage = "dummy message";
+            int port = 54320;
+            IPAddress multicastAddress = IPAddress.Parse("ff15::1:1");
```
Comment (tmds) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: Same here, reduce scope: `ff11`.

**Comment by tmds** [comment]
The tests were both using multicastInterface = 1, which is loopback, which isn't configured for ipv6 on most systems. So that explains the `Network is unreachable`.

I'll have a look at how a Linux system could be configured to deal with ipv6 multicast on loopback. In the meanwhile, can you make the proposed changes so we can have a look how they fail in CI?

**Comment by tmds** [comment]
> I'll have a look at how a Linux system could be configured to deal with ipv6 multicast on loopback.

Actually, since that will require special config on almost all system for this test to pass. Perhaps it would be better to try and find an interface that meets the criteria (ipv6, multicast) and use that instead.

**Comment by tmds** [comment]
> System.Net.Sockets.SocketException : Network is unreachable

This is the expected error due to loopback configuration.

> Waiting for received data timed out

This indicates we were able to send the data but not receive it.

> So far I have a similar C++ tests with my product and IPv6 multicast works fine in this Ubuntu machine.

I guess your test doesn't use loopback? Also, does it include receiving back the data? Are you using the same addresses/ports? Are you setting particular socket options?

**Comment by tmds** [comment]
@pepone can you change these lines:
https://github.com/dotnet/corefx/blob/fed2d5ff6842b2ce932c70162f3e06a8d0304030/src/Native/Unix/System.Native/pal_networking.c#L1375
https://github.com/dotnet/corefx/blob/fed2d5ff6842b2ce932c70162f3e06a8d0304030/src/Native/Unix/System.Native/pal_networking.c#L1413

IPPROTO_IP should be IPPROTO_IPV6

**Comment by pepone** [comment]
So Now it is working with Any Interface and still failing with loopback, this is consistent with C++/Java tests we have, where multicast doesn't work by default on Linux loopback, it should work fine with macOS and Windows.

**Comment by tmds** [comment]
If we'd want to run this test on Linux, we can check what network interfaces route ipv6 multicast, and then find out their interface index.
And use that instead of loopback interface:
```
$ route -n -A inet6 | grep 'ff00::/8'
ff00::/8                       ::                         U    256 17  1103 enp9s0
ff00::/8                       ::                         U    256 17   126 tun0
$ cat /sys/class/net/enp9s0/ifindex 
2
```

**Comment by pepone** [comment]
@tmds regarding remaining failures

UWP seems expected as it is currently disabled for IPv4 see https://github.com/pepone/corefx/blob/master/src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs#L102

RedHat.73.Amd64.Open-Release-x64 failure seems could be due to ipv6 firewall as mentioned in https://github.com/dotnet/corefx/pull/25552 I will add a similar fix

It is not clear what is the problem with macOS could this be a miss configuration? I will try to reproduce that locally

**Comment by wfurt** [comment]
Let me know if you need any help with OSX @pepone 
I can probably also help wth pushing firewall changes to our build infrastructure. 

In case of Fedora and RH, the default filter simply has no rules for UDP so it get's drop.
That is not not specific to multicast and I was wondering if we can detect that somehow. 
But I did not come to any solution I liked. Any thoughts @tmds ?

**Comment by pepone** [comment]
@wfurt Will be nice I you can try the test in macOS I suspect is a configuration issue, I have not been able to build corefx on my MacBook Air keeps failing with "Too many open files"

**Comment by tmds** [comment]
The mac test fails on the any interface the same way the linux test fails on the loopback interface: `No route to host`. So the mac CI machines probably don't route any IPv6 multicast.

I also think we are hitting the firewall on rhel7. It's unexpected this is passing on CentOS (they should behave the same).
@wfurt my way of detecting if it's the firewall: turn off the firewall and see if it starts to work :)

**Comment by wfurt** [comment]
Fine with me @tmds That may be doable for our official runs. But I would like to get to point when tests do run reasonably well for anybody. _if_ we can describe what exactly is needed  or _if_ can detect preconditions for tests we could get more reasonable outcome.

**Comment by tmds** [comment]
@wfurt totally agree

**Comment by tmds** [comment]
I think this PR looks good. Perhaps we can skip the OSes that fail and create 3 issues (Linux[Loopback], RHEL7[Any/Loopback] and OSX[Any/Loopback]) to investigate further?
The RHEL7 and Linux issues may be assigned to me.
@wfurt @stephentoub wdyt?

**Comment by pepone** [comment]
Should we also add [OuterLoop] anotation like IPv4 tests?

**Comment by wfurt** [comment]
yes, I think [Outerloop] would be appropriate @pepone

**Comment by tmds** [comment]
NETFX x86 failed with:
```
16:21:07       System.Net.Sockets.Tests.OSSupportTest.IOControl_SIOCATMARK_Success [FAIL]
16:21:07         Assert.Equal() Failure
16:21:07         Expected: 0
16:21:07         Actual:   1
16:21:07         Stack Trace:
16:21:07              at System.Net.Sockets.Tests.OSSupportTest.IOControl_SIOCATMARK_Success()
```
@dotnet-bot test NETFX x86 Release Build please

**Comment by tmds** [comment]
@dotnet-bot test Alpine.3.6 x64 Debug Build please

**Comment by tmds** [comment]
@wfurt @stephentoub ptal. For me this is good to merge. I will follow-up on the Linux[Loopback], RHEL7[Any/Loopback] test issues. 
Since we had to fix `SystemNative_SetIPv6MulticastOption` I think IPv6 multicast has not worked on .NET Core so far: this setter is used to subscribe to a multicast stream.

**Comment by tmds** [comment]
@wfurt @stephentoub ptal

**Comment by stephentoub** [comment]
@dotnet-bot test Alpine.3.6 x64 Debug Build please

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 6]
```diff
@@ -159,6 +159,81 @@ public void MulticastInterface_Set_InvalidIndex_Throws()
             }
         }
 
+        [OuterLoop] // TODO: Issue #11345
+        [Fact]
+        public async Task MulticastInterface_Set_IPv6_AnyInterface_Succeeds()
+        {
+            if (PlatformDetection.IsFedora || PlatformDetection.IsRedHatFamily7 || PlatformDetection.IsOSX)
+            {
+                return; // [ActiveIssue(24008)]
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: This issue is now closed.  Can this be re-enabled?

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 7]
```diff
@@ -159,6 +159,81 @@ public void MulticastInterface_Set_InvalidIndex_Throws()
...
+            string message = "hello";
+            int port;
+
+            using (Socket receiveSocket = CreateBoundUdpIPv6Socket(out port),
+                          sendSocket = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp))
+            {
+                receiveSocket.ReceiveTimeout = 1000;
+                receiveSocket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.AddMembership, new IPv6MulticastOption(multicastAddress, interfaceIndex));
+
+                // https://github.com/Microsoft/BashOnWindows/issues/990
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: Looks like this was supposed to be fixed a while ago.  Can this check be removed?

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop OSX x64 Release Build please
@dotnet-bot test Outerloop Linux x64 Debug Build please
@dotnet-bot test Outerloop Windows x64 Release Build please

**Comment by karelz** [comment]
The 3 failures in Outerloop Windows are all known issues
* `System.ServiceProcess.Tests.ServiceBaseTests` / `TestOnExecuteCustomCommand` - Win10 - Assert: pos 44
* `System.Net.Http.Functional.Tests.HttpProtocolTests` / `MaxResponseContentBufferSize_ThrowsIfTooSmallForContent` - Win7 - The handle is invalid
* `System.Net.Http.Functional.Tests.HttpProtocolTests` / `GetAsync_ExpectedStatusCodeAndReason_Success` - Win7 - The handle is invalid

**Comment by karelz** [comment]
@pepone can you please address the 2 outstanding feedback points from Stephen? Thanks! (I didn't notice them before merging :()

**Comment by pepone** [comment]
@karelz  should I push to this branch or use a separate pull request?

**Comment by karelz** [comment]
New PR please. Thank you!

**Pull request merged**